File: jetbrains-plugin/src/main/kotlin/com/ezevillo/gitreview/domain/SourcePreference.kt
package com.ezevillo.gitreview.domain

data class SourcePreferenceLevels(
    val workspaceValue: String? = null,
    val globalValue: String? = null,
)

private val VALID = setOf("remote", "local", "offline")

private fun asSource(value: String?): ReviewSource? {
    if (value != null && value in VALID) return ReviewSource.parse(value)
    return null
}

/** Workspace wins over user; default `"remote"`. */
fun resolveDefaultSource(levels: SourcePreferenceLevels): ReviewSource =
    asSource(levels.workspaceValue) ?: asSource(levels.globalValue) ?: ReviewSource.REMOTE

File: visualstudio-extension/src/GitReview.Domain/SourcePreference.cs
namespace GitReview.Domain;

public sealed record SourcePreferenceLevels(
    string? WorkspaceValue = null,
    string? GlobalValue = null);

public static class SourcePreference
{
    private static readonly HashSet<string> Valid = new(StringComparer.Ordinal)
    {
        "remote", "local", "offline",
    };

    private static ReviewSource? AsSource(string? value) =>
        value is not null && Valid.Contains(value) ? ReviewSourceExt.Parse(value) : null;

    /// <summary>Workspace wins over user; default remote.</summary>
    public static ReviewSource ResolveDefaultSource(SourcePreferenceLevels levels) =>
        AsSource(levels.WorkspaceValue) ?? AsSource(levels.GlobalValue) ?? ReviewSource.Remote;
}
